### ISSUE TITLE
Running grub-mkrescue command not building bootable ISO

### ISSUE THREAD

**Issue by abjrcode** [Microsoft/WSL#807]
Trying to use `grub-mkrescue` to create a bootable ISO image.
I am running: `grub-mkrescue -o myos.iso FOLDER_CONTAINING_OS_FILES`

The ISO file is created as expected but doesn't contain any grub information like the boot catalog or other grub files so it isn't bootable.

I expect to get a bootable ISO image. Running the same command against the same target folder from Ubuntu works and gives expected results.

My windows version is:  1607, build 14393.10

```
execve("/usr/bin/grub-mkrescue", ["grub-mkrescue", "-o", "os.iso", "isofiles/"], [/* 16 vars */]) = 0
brk(0)                                  = 0x2411000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6e10000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=22743, ...}) = 0
mmap(NULL, 22743, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f06a6e0a000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/liblzma.so.5", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320 \0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=137400, ...}) = 0
mmap(NULL, 2232456, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a67d0000
mprotect(0x7f06a67f1000, 2093056, PROT_NONE) = 0
mmap(0x7f06a69f0000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x20000) = 0x7f06a69f0000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260Z\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=228992, ...}) = 0
mmap(NULL, 2328416, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a6590000
mprotect(0x7f06a65c5000, 2093056, PROT_NONE) = 0
mmap(0x7f06a67c4000, 16384, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x34000) = 0x7f06a67c4000
mmap(0x7f06a67c8000, 1888, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f06a67c8000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0P \2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1840928, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6e00000
mmap(NULL, 3949248, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a61c0000
mprotect(0x7f06a637a000, 2097152, PROT_NONE) = 0
mmap(0x7f06a657a000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1ba000) = 0x7f06a657a000
mmap(0x7f06a6580000, 17088, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f06a6580000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320\16\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=14664, ...}) = 0
mmap(NULL, 2109744, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5fb0000
mprotect(0x7f06a5fb3000, 2093056, PROT_NONE) = 0
mmap(0x7f06a61b2000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2000) = 0x7f06a61b2000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libselinux.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0[\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=134296, ...}) = 0
mmap(NULL, 2238192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5d80000
mprotect(0x7f06a5da0000, 2093056, PROT_NONE) = 0
mmap(0x7f06a5f9f000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1f000) = 0x7f06a5f9f000
mmap(0x7f06a5fa1000, 5872, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f06a5fa1000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libudev.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0200\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=67600, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6df0000
mmap(NULL, 2164816, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5b60000
mprotect(0x7f06a5b70000, 2093056, PROT_NONE) = 0
mmap(0x7f06a5d6f000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xf000) = 0x7f06a5d6f000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libpcre.so.3", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260\27\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=252032, ...}) = 0
mmap(NULL, 2347200, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5920000
mprotect(0x7f06a595d000, 2093056, PROT_NONE) = 0
mmap(0x7f06a5b5c000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3c000) = 0x7f06a5b5c000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libcgmanager.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0206\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=108480, ...}) = 0
mmap(NULL, 2203688, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5700000
mprotect(0x7f06a5719000, 2093056, PROT_NONE) = 0
mmap(0x7f06a5918000, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x18000) = 0x7f06a5918000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libnih.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0`G\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=96280, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6de0000
mmap(NULL, 2191776, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a54e0000
mprotect(0x7f06a54f7000, 2093056, PROT_NONE) = 0
mmap(0x7f06a56f6000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x16000) = 0x7f06a56f6000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libnih-dbus.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\340,\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=38920, ...}) = 0
mmap(NULL, 2134040, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a52d0000
mprotect(0x7f06a52d8000, 2097152, PROT_NONE) = 0
mmap(0x7f06a54d8000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x8000) = 0x7f06a54d8000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libdbus-1.so.3", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@h\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=281552, ...}) = 0
mmap(NULL, 2377408, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a5080000
mprotect(0x7f06a50c4000, 2093056, PROT_NONE) = 0
mmap(0x7f06a52c3000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x43000) = 0x7f06a52c3000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/librt.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0P#\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=31792, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6dd0000
mmap(NULL, 2129016, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a4e70000
mprotect(0x7f06a4e77000, 2093056, PROT_NONE) = 0
mmap(0x7f06a5076000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6000) = 0x7f06a5076000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0po\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=141574, ...}) = 0
mmap(NULL, 2217264, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f06a4c50000
mprotect(0x7f06a4c69000, 2093056, PROT_NONE) = 0
mmap(0x7f06a4e68000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x18000) = 0x7f06a4e68000
mmap(0x7f06a4e6a000, 13616, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f06a4e6a000
close(3)                                = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6dc0000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6db0000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6da0000
arch_prctl(ARCH_SET_FS, 0x7f06a6da0840) = 0
mprotect(0x7f06a657a000, 16384, PROT_READ) = 0
mprotect(0x7f06a4e68000, 4096, PROT_READ) = 0
mprotect(0x7f06a5076000, 4096, PROT_READ) = 0
mprotect(0x7f06a52c3000, 4096, PROT_READ) = 0
mprotect(0x7f06a56f6000, 4096, PROT_READ) = 0
mprotect(0x7f06a54d8000, 4096, PROT_READ) = 0
mprotect(0x7f06a5918000, 8192, PROT_READ) = 0
mprotect(0x7f06a5b5c000, 4096, PROT_READ) = 0
mprotect(0x7f06a5d6f000, 4096, PROT_READ) = 0
mprotect(0x7f06a61b2000, 4096, PROT_READ) = 0
mprotect(0x7f06a5f9f000, 4096, PROT_READ) = 0
mprotect(0x7f06a67c4000, 4096, PROT_READ) = 0
mprotect(0x7f06a69f0000, 4096, PROT_READ) = 0
mprotect(0x6b8000, 4096, PROT_READ)     = 0
mprotect(0x7f06a6c22000, 4096, PROT_READ) = 0
munmap(0x7f06a6e0a000, 22743)           = 0
set_tid_address(0x7f06a6da0b10)         = 707
set_robust_list(0x7f06a6da0b20, 24)     = 0
futex(0x7fffe2acbe60, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 1, NULL, 7f06a6da0840) = -1 EAGAIN (Resource temporarily unavailable)
rt_sigaction(SIGRTMIN, {0x7f06a4c569f0, [], SA_RESTORER|SA_SIGINFO, 0x7f06a4c60330}, NULL, 8) = 0
rt_sigaction(SIGRT_1, {0x7f06a4c56a80, [], SA_RESTORER|SA_RESTART|SA_SIGINFO, 0x7f06a4c60330}, NULL, 8) = 0
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
getrlimit(RLIMIT_STACK, {rlim_cur=8192*1024, rlim_max=8192*1024}) = 0
statfs("/sys/fs/selinux", 0x7fffe2acbe30) = -1 ENOENT (No such file or directory)
statfs("/selinux", 0x7fffe2acbe30)      = -1 ENOENT (No such file or directory)
brk(0)                                  = 0x2411000
brk(0x2432000)                          = 0x2432000
open("/proc/filesystems", O_RDONLY)     = 3
fstat(3, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6d90000
read(3, "nodev      sysfs\nnodev      root"..., 1024) = 367
read(3, "", 1024)                       = 0
close(3)                                = 0
munmap(0x7f06a6d90000, 4096)            = 0
open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2919792, ...}) = 0
mmap(NULL, 2919792, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f06a4987000
close(3)                                = 0
gettimeofday({1470602047, 783029}, NULL) = 0
mkdir("/tmp/grub.niI5tG", 0700)         = 0
mkdir("", 0755)                         = -1 ENOENT (No such file or directory)
mkdir("/tmp", 0755)                     = -1 EEXIST (File exists)
mkdir("/tmp/grub.niI5tG", 0755)         = -1 EEXIST (File exists)
mkdir("/tmp/grub.niI5tG/boot", 0755)    = 0
mkdir("/tmp/grub.niI5tG/boot/grub", 0755) = 0
mkdir("/tmp/grub.niI5tG/boot/grub/roms", 0755) = 0
stat("/usr/lib/grub/i386-pc", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-efi", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-qemu", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-coreboot", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-multiboot", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-ieee1275", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/x86_64-efi", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/mipsel-loongson", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/sparc64-ieee1275", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/powerpc-ieee1275", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/mipsel-arc", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/mips-arc", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/ia64-efi", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/arm-uboot", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/arm-efi", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/mipsel-qemu_mips", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/mips-qemu_mips", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/i386-xen", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/x86_64-xen", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
stat("/usr/lib/grub/arm64-efi", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
time(NULL)                              = 1470602047
open("/etc/localtime", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=1073, ...}) = 0
fstat(3, {st_mode=S_IFREG|0644, st_size=1073, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06a6d90000
read(3, "TZif2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\2\0\0\0\2\0\0\0\0"..., 512) = 512
lseek(3, -124, SEEK_CUR)                = 388
read(3, "TZif2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\2\0\0\0\2\0\0\0\0"..., 512) = 512
read(3, "z\260\246\20\0\0\0\0{\316\273\220\0\0\0\0|\231\302\220\0\0\0\0}\256\235\220\0\0\0\0"..., 512) = 173
close(3)                                = 0
munmap(0x7f06a6d90000, 4096)            = 0
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f06a6da0b10) = 708
wait4(708, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 708
openat(AT_FDCWD, "/tmp/grub.niI5tG", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, /* 3 entries */, 32768)     = 67
lstat("/tmp/grub.niI5tG/boot", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
openat(AT_FDCWD, "/tmp/grub.niI5tG/boot", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
getdents(4, /* 3 entries */, 32768)     = 67
lstat("/tmp/grub.niI5tG/boot/grub", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot/grub", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot/grub", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
openat(AT_FDCWD, "/tmp/grub.niI5tG/boot/grub", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 5
getdents(5, /* 3 entries */, 32768)     = 67
lstat("/tmp/grub.niI5tG/boot/grub/roms", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot/grub/roms", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
stat("/tmp/grub.niI5tG/boot/grub/roms", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
openat(AT_FDCWD, "/tmp/grub.niI5tG/boot/grub/roms", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 6
brk(0x2455000)                          = 0x2455000
getdents(6, /* 2 entries */, 32768)     = 43
getdents(6, /* 0 entries */, 32768)     = 0
rmdir("/tmp/grub.niI5tG/boot/grub/roms") = 0
brk(0x244d000)                          = 0x244d000
close(6)                                = 0
getdents(5, /* 0 entries */, 32768)     = 0
rmdir("/tmp/grub.niI5tG/boot/grub")     = 0
brk(0x2445000)                          = 0x2445000
close(5)                                = 0
getdents(4, /* 0 entries */, 32768)     = 0
rmdir("/tmp/grub.niI5tG/boot")          = 0
brk(0x243d000)                          = 0x243d000
close(4)                                = 0
getdents(3, /* 0 entries */, 32768)     = 0
rmdir("/tmp/grub.niI5tG")               = 0
brk(0x2435000)                          = 0x2435000
close(3)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

From the above, it seems that `/usr/lib/grub` folder is not there which probably means grub isn't installed.

Can this be done, what do I need to do to make it work ?
Thank you

**Comment by fpqc** [comment]
I sort of think that this wouldn't work because grub uses a lot of low-level stuff, and making a rescue image might pull stuff out of the existing OS, but I could be wrong.  If you notice, all of the `/usr/lib/grub-$ARCH` directories are empty, and maybe that's why?  Also, I don't know if grub tries to mount some kind of image before writing to it, but if it does, it won't work, since wsl doesn't support mounting anything on the linux side at all.  A few versions back, you could mount a vhdx formatted with ntfs in Windows as a directory and have it available in `/mnt/physicaldriveletter/.../mountpoint`, but the team seems to have removed it (although we're not sure about the reasons atm).

**Comment by abjrcode** [comment]
Yeah I figured it would be low level task, hoping we get a comment from someone in the team to see if this is somehow _(will be)_ supported.

**Comment by benhillis** [comment]
I think @fpqc might be right that the tools is attempting to use grub binaries from the host OS (which we do not have because we do not have a boot loader).  I suspect this would also fail in a a docker container using the [same image our filesystem is based off of](http://uec-images.ubuntu.com/releases/14.04/release/ubuntu-14.04-server-cloudimg-amd64-root.tar.gz).

**Comment by AlexandreRouma** [comment]
UPDATE ! I can confirm that by copying the folder `/usr/lib/grub/i386-pc`, `grub-mkrescue` successfully generates a bootable ISO !

**Comment by DarthSpock** [comment]
@AlexandreRouma could you provide repo steps showing this? Also can you post a test result of the iso being successfully installed using both WSL and actual Ubuntu? I would like to ensure it's a clean image that's generated.

**Comment by AlexandreRouma** [comment]
Yeah sure !

First of all, here are the version of everything involved:
Source OS: Ubuntu 16.04 LTS installed in Virtual Box Version 5.1.30 r118389 (Qt5.6.2)
LSW: Linux version 4.4.0-43-Microsoft

Here is how to replicate:

1. In LSW: run `apt-get install grub-common`
2. In the source OS: Copy `/usr/lib/grub/i386-pc/`
3. In LSW: Paste i386-pc into `/usr/lib/grub/`
4. In LSW: Run the desired `grub-mkrescue` command. In my case: `grub-mkrescue -o MemeOS.iso ISO` (The contents of `ISO` can be found on my github (MemeOS) )

The generated files from LSW and Ubuntu will be added as attachment inside a ZIP file.

Both ISOs are confirmed working on real hardware, QEMU, BOCHS and Virtual Box.

[ISOs.zip](https://github.com/Microsoft/WSL/files/1620048/ISOs.zip)

**Comment by DarthSpock** [comment]
Thanks! I think the use case I'm thinking may differ slightly than yours. I'm actually looking at seeing about installing the iso within the WSL environment, starting with for test purposes if I could say take the Arch or Kubuntu iso and install it within WSL Ubuntu and upon installation complete, have a fully ready to go OS using VcxSrv for the GUI portions.  Only toying with this idea as there are no 18.04 available repositories for adding different distros such as Kubuntu or Arch. There are iso files available though I was able to download. I may tweak some things about the OS at a later time but right now, I'm trying to see if the capability is there.

**Comment by therealkenc** [comment]
> UPDATE ! I can confirm that by copying the folder /usr/lib/grub/i386-pc, grub-mkrescue successfully generates a bootable ISO !

That's great; thanks for the update. Open question is, has this worked all along. In the OP the fail was just:
```
stat("/usr/lib/grub/i386-pc", 0x7fffe2acba30) = -1 ENOENT (No such file or directory)
```
Which would fail all the same on Real Linux™ if that directory didn't exist. Seems doubtful anything was actually fixed since August 2016.

> but do you think it would be possible to have the option to install (even if they're not used) GRUB binaries ?

That's a good question. I've re-opened related #1965. As mentioned over there, this would require stubbing enough to make the installer happy, even if `grub` itself can't run.

**Comment by therealkenc** [comment]
Another option, which is pretty reasonable in the circumstance, would be to upstream a fix (call it a work-around if you prefer) to the installer itself to be a little more tolerable of its surroundings. I can even contrive a Real Linux use case. On Real Linux there is no guarantee there is a block device for the rootfs either, because it might be a diskless workstation. If this were 1994, anyway. On such a machine `grub-mkrescue` ought to be able to be installed too.

Wonder what is actually causing the fail. No one ever did an `strace` over in #1965 ([message](https://github.com/Microsoft/WSL/issues/1965#issuecomment-296340575)), but I still suspect it is `libblkid-dev` related.

**Comment by AlexandreRouma** [comment]
Yeah, it would be much better, because copying files from one machine to another is a bit sketchy 😐

**Comment by zunderscore** [comment]
I've been fighting with this same issue today, but I found a viable workaround that's a little better than copying from another system. :)

If you install the `grub-pc-bin` package, then `grub-mkrescue` will actually create a bootable ISO! It looks like it adds the required files to both `/usr/lib/grub/i386-pc` and `/usr/lib/grub/x86_64-pc`. I've tested this both on FCU (16299) and the latest fast ring (17074). I know this doesn't necessarily solve the underlying issue of the binaries missing to begin with, but it seems like a very small workaround. Hope this helps!
